New Blog Post: Setting Up GitHub Actions for V (#20)
vlang/vlang-blog: content/blog/setting-up-github-actions-for-v.md

page: Blog
authorname: Subhomoy Haldar
authorlink: subhomoy-haldar
categories: ["tutorials"]
date: 2022-03-14T12:28:35+05:30
image:
---

## The Article Series

> 1. [The Complete Beginner’s Guide to CLI Apps in V](/the-complete-beginners-guide-to-cli-apps-in-v/)
> 2. [Elevate Your V Project With Unit Tests](/elevate-your-v-project-with-unit-tests/)
> 3. **Setting Up GitHub Actions for V** _(You are here)_

## What is CI/CD, DevOps, etc.?

Software Development is a very involved process, with a lot of moving parts.
This is especially true when you have several developers working in one or
more teams. The ultimate objective of these endeavours is to understand the
client requirements, make sure everybody is on the same page, and eventually,
deliver the final product.

**Continuous Integration** for the most part is about the second item in the
list: _making sure everyone is on the same page_. Individuality is important

> for problem solving: it offers new angles, allows people to shine where they
> are good at. But it can be a problem in case of things like formatting

Reviewer: ```suggestion
for problem solving: it offers new angles and allows people to shine where they
are able. But it can be a problem in case of things like formatting
```

preferences and build environments. Homogeneity is preferred in these
instances. Therefore, the decision makers band together to lay down some
ground rules for everyone to follow. These include things such as:

1. Expected behaviour that is standardized using _tests_.
2. _Source Code formatting_ guidelines that are strictly enforced.

> 3. Managing build and release _environments_ with relevant dependencies managed.

Reviewer: ```suggestion
3. Managing build and release _environments_ with relevant dependencies.
```

resource intensive. You would have to setup a CI server, install the
relevant software on it, setup the notification systems, and put a lot
of time into general infrastructure management.

Now it's no longer the case. Several providers like
[GitLab](https://docs.gitlab.com/ee/ci/index.html) and
[GitHub](https://docs.github.com/en/actions)
provide first-party CI/CD services. They offer generous free-tiers which
exceed the requirements of regular developers and small businesses.

If you put in a bit of elbow grease, you can set up your own CI/CD service.
For example, a popular FOSS pairing is
[Woodpecker CI](https://woodpecker-ci.org/) with
[Gitea](https://gitea.io/en-us/).

You still have to write the pipelines for all your repositories, but
getting started with CI for a project is much easier than it used to be.

## Introduction to GitHub Actions

We will stick to [GitHub Actions](https://docs.github.com/en/actions) for
this tutorial because it is the most widely used CI/CD service in the
community at present. We can follow up with instructions for other CI services
if there's demand for it.

Let us start by discussing some terminology.

> A _workflow_ is a set of instructions described in [YAML](https://yaml.org/).

Reviewer 2: ```suggestion
A _workflow_ is a set of jobs, plus a specification on when to run them, described in [YAML](https://yaml.org/).
```

Author: I'm sorry I read the suggestion the other way around.

We define one workflow per YAML file. All workflows are stored in the
`.github/workflows` directory. They are generally triggered on every push,
but they can be trigged manually, or on every tag, or periodically.

There can be several _jobs_ per _workflow_. Each job is a set of steps
executed sequentially. Every _step_ is either a shell script that will
be run, or an _action_ from the marketplace. Each job is run on the same

> _runner_ (think: machine), so the data is generally persisted across the steps.

Reviewer: ```suggestion
_runner_ (think: server instance - may be a physical machine, a cloud instance, etc.), so the data is generally persisted across the steps.
```

_ubuntu_ runners. So it might be worthwhile to quickly prototype on
`ubuntu-latest` and then add `macos-latest` and/or `windows-latest` when
needed.

We can now defined the _steps_ for the `build` job:

```yaml {tabWidth=2}
# ...

jobs:
  build:
    strategy:
      matrix:
        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
    runs-on: ${{ matrix.os }}

    steps:
      - name: Install V
        uses: vlang/setup-v@v1
        with:
          check-latest: true

      - name: Checkout ${{ github.event.repository.name }}
        uses: actions/checkout@v2

      - name: Check if code is formatted
        run: |
          v fmt -diff .
          v fmt -verify .

      - name: Build ${{ github.event.repository.name }}
        run: v .

      - name: Run Tests
        run: v test .
```

First, we use the `vlang/setup-v@1` GitHub Action to install V for our runner.
We set `check-latest` to `true` to get the latest updates. Leave this out if
you prefer more stability. Then we clone the current project using the
`actions/checkout@v2` action.

We can now start checking the codebase to ensure that it passes the minimum

> standards we set. V has an opinionated auto-formatter available called `vfmt`.

Reviewer 2: ```suggestion
standards we set. V ships with its own opinionated tool for formatting V source code.
```

Here is a sample usage:

```bash
# Format all V files recursively in this directory
# AND rewrite them
v fmt -w .
# ^^^
# You'll most likely use this all the time from the root of
# your V projects.

# Display the formatted output of a sample V file to stdout
v fmt sample.v
```

You can run `v help fmt` to get more details on all the available options.
Two more options we have at our disposal are:

1. `v fmt -diff .` which displays the difference between the original and

>    formatted source, if any. Ideally, we don't any difference before we

Reviewer: ```suggestion
   formatted source, if any. Ideally, we don't want any difference before we
```

   push the commits.
2. `v fmt -verify .` which reports an error if it finds an unformatted
   file in the repository.

Note that we do not automatically format the code in the CI because it might
lead to huge `diff`s later on and a lot of potentially redundant commits. It
is also easier to have our tests be _read-only_.

> This step executes quickly. Next we make sure that we can build our project.

Reviewer 2: what step executes quickly?

Reviewer 2: At this point, there was a detour of 2-3 paragraphs that discuss `v fmt`. It would be helpful to name the step explicitly, instead of referring to it with `this`.

can build our project. This is as simple as running the `v .` command. This
will build the project fast. If you plan to distribute the built executable,
it is recommended to use `v -prod .` to generate an optimised executable,
which takes slightly longer. You can also choose to generate optimised `-prod`
builds, only when a tag is pushed, instead of on each commit or Pull Request.

> Finally, we run the unit tests (if any) to make sure that we have not
> regressed, in the form of failing tests. V's in-built testing framework

Reviewer: ```suggestion
Finally, we run the unit tests (if any) to make sure that we have not
regressed. This will be shown by failing tests. V's built-in testing framework
```
FYI... `in-built` is often used to mean the same as `built-in`, but `built-in` is much more popular usage (`in-built` is mostly only use in the UK).  Some consider them to be slightly different meaning... `in-build` meaning more of an innate piece of the original (like your heart and lungs), while `built-in` is extra parts added to the original (cyborgs?  :-)).

If you want to test a pre-existing workflow (say, after cloning a project),
you might need to enable the workflows in the _Actions_ tab of the
repository. Then push a commit to test.

For testing purposes, we can push an empty commit to trigger the CI:

```bash
git commit -m "trigget CI attempt 1" --allow-empty
```

We can keep iterating, making tweaks, committing more changes, until the
workflow passes successfully. Any changes we make to the code from now on
will always be checked by the CI. If the workflow fails, we'll get an
email notification from GitHub.

## A Sophisticated Example

For now, we have only talked about Continuous Integration. As mentioned
before, we can run a separate workflow when git tags are pushed to the
remote repository. In this repository, we can do the same things as in
the regular workflow, but include some novel things things as building
an optimised executable and uploading the executable built for all
available runners for later use. The latter would be a good example of
Continuous Delivery (CD). If we add steps for authentication and
deployment of the executable, then that would be Continuous Deployment
(also CD).

> An example of such an workflow can be found in the author's

Reviewer: ```suggestion
An example of such a workflow can be found in the author's
```

[klonol](https://github.com/hungrybluedev/klonol/tree/main/.github/workflows)
project.

## Concluding Thoughts

A CI/CD setup is a great way to automate a lot of tedious work. It allows
the developers to focus on fixing bugs, adding features, and they do not

> have to worry about regression, or stepping on each others' toes.

Reviewer 2: ```suggestion
have to worry about finding regression, or stepping on each others' toes.
```

## Testing the CI with sample commits

The final workflow file will be as follows:

```yaml {linenos=table}
name: CI

on:
  push:
    branches: [main]

jobs:
  build:
    strategy:
      matrix:
        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
    runs-on: ${{ matrix.os }}

    steps:
      - name: Install V
        uses: vlang/setup-v@v1
        with:
          check-latest: true

      - name: Checkout ${{ github.event.repository.name }}
        uses: actions/checkout@v2

      - name: Check if code is formatted
        run: |
          v fmt -diff .
          v fmt -verify .
      - name: Build ${{ github.event.repository.name }}
        run: v .

      - name: Run Tests
        run: v test .
```

All V projects are initialised with git. Therefore, all we need to do to
trigger a CI run is to make sure that the remote is pointed to a repository
on GitHub, add the `ci.yml` file, commit the changes and push.

If you want to test a pre-existing workflow (say, after cloning a project),
you might need to enable the workflows in the _Actions_ tab of the
repository. Then push a commit to test.

For testing purposes, we can push an empty commit to trigger the CI:

```bash

> git commit -m "trigget CI attempt 1" --allow-empty

Reviewer 3: trigget → trigger?

6. Run the unit tests if any.
7. Optionally upload the built artifacts to be used later.

> Artifacts are files that are produced by the build process. They can be
> executables, logs, results, trained models, or any other file that has a
> significance to the end user, and can be reused after the pipeline has
> finished.
>
> Refer to the [documentation](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts)
> for more information.

The reference workflow files are available here:
[workflows](https://github.com/hungrybluedev/geo/tree/main/.github/workflows).

## Setting Up Github Actions CI for V

First we create a new file `ci.yml` in the `.github/workflows` directory.

> We have a few options to choose from:

Reviewer: Hmm... not sure this section is needed - it seems incongruously basic, compared to the rest of the article.

Author: I wrote this in response to this question:

> How does one get to the .github/workflows directory, to create a new file? I did it by using the "Actions" menu on the github page, and then configuring their Simple workflow. Is there another way I don't know about?

Maybe I can shorten the explanation here.

it on every tag, or on pushes to a pull request.

Next, we define the primary `build` job in the list of jobs. We define an
array called `os` in the _build matrix_ with our desired platforms. This goes
in the `strategy` section of the job. More details about build matrices can be
found in the
[official documentation](https://docs.github.com/en/actions/using-jobs/using-a-build-matrix-for-your-jobs).
Briefly, they help us reuse the same steps for different platforms by performing
substitution of variables. For example, `${{ matrix.os }}` will be replaced by

> all the OSes we defined in the matrix, and the job will be run for each of them.

Reviewer: ```suggestion
each OS we defined in the matrix, one at a time, and the job will be run for each of them.
```
